drivers/spi/spi_eos_s3.c

```diff
@@ -0,0 +1,353 @@
+	regs->SSIENR = SSIENR_SSI_DISABLE;
+
+	/* Disable Slave Select */
+	regs->SER = 0;
+
+	/* Find divisor to setup clock rate */
+	uint32_t clock_div = SPI_CLK / config->frequency;
+
+	/* Divisor must be an even number */
+	if (clock_div % 2) {
+		clock_div++;
+	}
+
+	/* This driver doesn't work correctly with clock rate higher than 2MHz.
+	 * 10MHz / 6 ~= 1.6 MHz
+	 */
+	if (clock_div < 6)
+		return -EINVAL;
+
+	regs->BAUDR = clock_div;
+
+	ctrlr0 = SPI_DATASIZE_8BIT;
+	ctrlr0 |= (config->operation & SPI_MODE_CPOL) ? SPI_POLARITY_HIGH
+		: SPI_POLARITY_LOW;
+	ctrlr0 |= (config->operation & SPI_MODE_CPHA) ? SPI_PHASE_1EDGE
+		: SPI_PHASE_2EDGE;
+
+	switch (mode) {
+	case TXRX_MODE:
+		ctrlr0 |= CTRLR0_TMOD_TX_RX;
+		break;
+	case TX_MODE:
+		ctrlr0 |= CTRLR0_TMOD_TX;
+		break;
+	case RX_MODE:
+		ctrlr0 |= CTRLR0_TMOD_RX;
+		/* Set number of data frames to read */
+		break;
+	case EEPROM_READ_MODE:
+		/* Not supported */
+		break;
+	}
+
+	regs->CTRLR0 = ctrlr0;
+
+	/* Choose Slave Select line */
+	regs->SER = BIT(config->slave);
+
+	/* Enable SPI controller */
+	regs->SSIENR = SSIENR_SSI_EN;
+
+	return 0;
+}
+
+static void spi_eos_s3_set_nframes(SPI_TypeDef *regs, uint16_t nfrm)
+{
+	/* Disable SPI controller */
+	regs->SSIENR = SSIENR_SSI_DISABLE;
+
+	/* Program number of frames to receive */
+	uint16_t ctrlr1 = nfrm - 1;
+
+	regs->CTRLR1 = ctrlr1;
+
+	/* Enable SPI controller */
+	regs->SSIENR = SSIENR_SSI_EN;
+}
+
+/* Finish any ongoing writes and drop any remaining read data */
+static void spi_eos_s3_finish(SPI_TypeDef *regs)
+{
+	while (!(regs->SR & SR_TFE)) {
+	}
+
+	while (regs->SR & SR_RFNE) {
+		(void)regs->DR0;
+	}
+}
+
+static void spi_eos_s3_tx(SPI_TypeDef *regs, uint8_t frame)
+{
+	uint8_t c = 0;
+	/* Wait for not full TX FIFO */
+	while (!(regs->SR & SR_TFNF)) {
+	}
+
+	c = frame;
+	regs->DR0 = c;
+}
+
+static int spi_eos_s3_rx(SPI_TypeDef *regs)
+{
+	if (!(regs->SR & SR_RFNE)) {
+		return NO_RX_DATA;
+	}
+
+	return regs->DR0;
+}
+
+static int spi_eos_s3_init(const struct device *dev)
+{
+	int ret;
+	const struct spi_eos_s3_config *config = dev->config;
+
+	ret = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline int spi_eos_s3_next_tx(struct spi_eos_s3_data *data)
+{
+	int tx_frame;
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
+	} else {
+		tx_frame = NO_TX_DATA;
+	}
+
+	return tx_frame;
+}
+
+static int spi_eos_s3_shift_frames(SPI_TypeDef *regs,
+				   struct spi_eos_s3_data *data)
+{
+	int tx_frame;
+	int rx_frame;
+
+	/* Do not send data when continuous RX mode is selected */
+	if (data->mode != RX_MODE) {
+		tx_frame = spi_eos_s3_next_tx(data);
+		/* Do not send redundant data */
+		if (tx_frame != NO_TX_DATA) {
+			spi_eos_s3_tx(regs, (uint8_t)tx_frame);
+			spi_context_update_tx(&data->ctx, 1, 1);
+		}
+	}
+
+	rx_frame = spi_eos_s3_rx(regs);
+
+	if (rx_frame == NO_RX_DATA)
```

Comment:
even one liner if statements need {} per-zephyr code style (apply that where relevant)